In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from pydataset import data

# 1. Attendance Data
Load the `attendance.csv` file and calculate an attendance percentage for each student. 
- One half day is worth 50% of a full day, and 
- 10 tardies is equal to one absence.

In [2]:
attendance = pd.read_csv('untidy-data/attendance.csv', encoding = 'unicode escape')

In [143]:
attendance

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [4]:
#melted the multiple columns into one col:date and product into another col:attendance
attendance_melt = attendance.melt(id_vars = 'Unnamed: 0',
                                 var_name = 'date', value_name = 'attendance')
attendance_melt

Unnamed: 0        date attendance
0       Sally  2018-01-01          P
1        Jane  2018-01-01          A
2       Billy  2018-01-01          A
3        John  2018-01-01          P
4       Sally  2018-01-02          T
5        Jane  2018-01-02          P
6       Billy  2018-01-02          T
7        John  2018-01-02          T
8       Sally  2018-01-03          T
9        Jane  2018-01-03          T
10      Billy  2018-01-03          A
11       John  2018-01-03          H
12      Sally  2018-01-04          H
13       Jane  2018-01-04          T
14      Billy  2018-01-04          A
15       John  2018-01-04          P
16      Sally  2018-01-05          P
17       Jane  2018-01-05          T
18      Billy  2018-01-05          H
19       John  2018-01-05          P
20      Sally  2018-01-06          A
21       Jane  2018-01-06          T
22      Billy  2018-01-06          T
23       John  2018-01-06          T
24      Sally  2018-01-07          T
25       Jane  2018-01-07          A
26      Billy  2018-01-07          P
27       John  2018-01-07          P
28      Sally  2018-01-08          T
29       Jane  2018-01-08          T
30      Billy  2018-01-08          T
31       John  2018-01-08          P

In [5]:
#renamed first col to a suitable name
attendance_melt.rename(columns = {'Unnamed: 0' : 'name'}, inplace = True)
attendance_melt.head()

name        date attendance
0  Sally  2018-01-01          P
1   Jane  2018-01-01          A
2  Billy  2018-01-01          A
3   John  2018-01-01          P
4  Sally  2018-01-02          T

In [139]:
#attendance_melt.drop(columns = 'date', inplace = True)
#attendance_melt

In [140]:
# replaced attendance variable to numerical values
# Present=1, Absent=0, Halfday=0.5, Tardy=.9

#attendance_melt.replace((to_replace ="P", value = 1), (to_replace = 'T', value =  0.9))
attendance_melt = attendance_melt.replace(to_replace ="P", value = 1).replace(to_replace ="T", value = 0.9).replace(to_replace ="A", value = 0).replace(to_replace ="H", value = 0.5)
attendance_melt.head(12)

name        date  attendance
0   Sally  2018-01-01         1.0
1    Jane  2018-01-01         0.0
2   Billy  2018-01-01         0.0
3    John  2018-01-01         1.0
4   Sally  2018-01-02         0.9
5    Jane  2018-01-02         1.0
6   Billy  2018-01-02         0.9
7    John  2018-01-02         0.9
8   Sally  2018-01-03         0.9
9    Jane  2018-01-03         0.9
10  Billy  2018-01-03         0.0
11   John  2018-01-03         0.5

In [25]:
#created an avg of each student's attendance over the 8 days
attendance_melt.groupby('name').mean('attendance')

attendance
name             
Billy      0.5250
Jane       0.6875
John       0.9125
Sally      0.7625

In [167]:
attendance_melt.groupby('name').attendance.mean()

name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: attendance, dtype: float64

In [141]:
#attendance_tidy = attendance_melt.pivot_table(index= 'date', 
#                                              columns = 'name', values = 'attendance')

In [142]:
#attendance_tidy

In [11]:
#attendance_tidy.shape

(8, 4)

In [162]:
#attendance_tidy.T

#### A different take with same results

In [144]:
attendance.rename(columns = {'Unnamed: 0' : 'name'}, inplace = True)
attendance.head()

name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0  Sally          P          T          T          H          P          A   
1   Jane          A          P          T          T          T          T   
2  Billy          A          T          A          A          H          T   
3   John          P          T          H          P          P          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T  
2          P          T  
3          P          P

In [146]:
# replace using a dictionary
att = attendance.replace({'P': 1.0, 'A': 0.0, 'T': 0.9, 'H': 0.5})
att

name  2018-01-01  2018-01-02  2018-01-03  2018-01-04  2018-01-05  \
0  Sally         1.0         0.9         0.9         0.5         1.0   
1   Jane         0.0         1.0         0.9         0.9         0.9   
2  Billy         0.0         0.9         0.0         0.0         0.5   
3   John         1.0         0.9         0.5         1.0         1.0   

   2018-01-06  2018-01-07  2018-01-08  
0         0.0         0.9         0.9  
1         0.9         0.0         0.9  
2         0.9         1.0         0.9  
3         0.9         1.0         1.0

In [160]:
# changed 'name' col to be my index
att_tidy = att.pivot_table(index = 'name')
att_tidy

2018-01-01  2018-01-02  2018-01-03  2018-01-04  2018-01-05  2018-01-06  \
name                                                                            
Billy         0.0         0.9         0.0         0.0         0.5         0.9   
Jane          0.0         1.0         0.9         0.9         0.9         0.9   
John          1.0         0.9         0.5         1.0         1.0         0.9   
Sally         1.0         0.9         0.9         0.5         1.0         0.0   

       2018-01-07  2018-01-08  
name                           
Billy         1.0         0.9  
Jane          0.0         0.9  
John          1.0         1.0  
Sally         0.9         0.9

In [161]:
att_tidy.mean(axis=1)


name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
dtype: float64

# 2. Coffee Levels

2a. Read the `coffee_levels.csv` file.

In [26]:
coffee_levels = pd.read_csv('untidy-data/coffee_levels.csv')
coffee_levels.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

In [27]:
coffee_levels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   hour           30 non-null     int64  
 1   coffee_carafe  30 non-null     object 
 2   coffee_amount  30 non-null     float64
dtypes: float64(1), int64(1), object(1)
memory usage: 848.0+ bytes


In [29]:
coffee_levels

hour coffee_carafe  coffee_amount
0      8             x       0.816164
1      9             x       0.451018
2     10             x       0.843279
3     11             x       0.335533
4     12             x       0.898291
5     13             x       0.310711
6     14             x       0.507288
7     15             x       0.215043
8     16             x       0.183891
9     17             x       0.391560
10     8             y       0.189297
11     9             y       0.521502
12    10             y       0.023163
13    11             y       0.235529
14    12             y       0.017009
15    13             y       0.997464
16    14             y       0.058361
17    15             y       0.144644
18    16             y       0.544676
19    17             y       0.594126
20     8             z       0.999264
21     9             z       0.915990
22    10             z       0.144928
23    11             z       0.311495
24    12             z       0.771947
25    13             z       0.398520
26    14             z       0.864464
27    15             z       0.436364
28    16             z       0.280621
29    17             z       0.436677

2b. Transform the data so that each carafe is in it's own **column**.

In [30]:
coffee_levels.head(2)

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018

In [163]:
coffee_pivot = coffee_levels.pivot_table(index = 'hour', 
                         columns = 'coffee_carafe', values = 'coffee_amount')
coffee_pivot

coffee_carafe         x         y         z
hour                                       
8              0.816164  0.189297  0.999264
9              0.451018  0.521502  0.915990
10             0.843279  0.023163  0.144928
11             0.335533  0.235529  0.311495
12             0.898291  0.017009  0.771947
13             0.310711  0.997464  0.398520
14             0.507288  0.058361  0.864464
15             0.215043  0.144644  0.436364
16             0.183891  0.544676  0.280621
17             0.391560  0.594126  0.436677

2c. Is this the best shape for the data?

In [164]:
# I think this this looks to be neater. Each carafe is used on a different day.

# 3. Cake Recipes

3a. Read the `cake_recipes.csv` data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [33]:
cake_recipes = pd.read_csv('untidy-data/cake_recipes.csv')
cake_recipes.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

In [34]:
cake_recipes.shape

(8, 5)

In [35]:
cake_recipes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   recipe:position  8 non-null      object 
 1   225              8 non-null      float64
 2   250              8 non-null      float64
 3   275              8 non-null      float64
 4   300              8 non-null      float64
dtypes: float64(4), object(1)
memory usage: 448.0+ bytes


In [36]:
cake_recipes

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

3b. Tidy the data as necessary.

In [42]:
#rename 1st col to remove ':'
cake_recipes = cake_recipes.rename(columns = {'recipe:position' : 'recipe_position'})

In [43]:
cake_recipes.head()

recipe_position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

In [44]:
#create 2 new cols by splitting col[0]
cake_recipes[['recipe_name', 'oven_position']] = cake_recipes.recipe_position.str.split(':', expand = True)

In [46]:
cake_recipes.head()

recipe_position        225        250        275        300 recipe_name  \
0        a:bottom  61.738655  53.912627  74.414730  98.786784           a   
1           a:top  51.709751  52.009735  68.576858  50.228470           a   
2        b:bottom  57.095320  61.904369  61.196980  99.248541           b   
3           b:top  82.455004  95.224151  98.594881  58.169349           b   
4        c:bottom  96.470207  52.001358  92.893227  65.473084           c   

  oven_position  
0        bottom  
1           top  
2        bottom  
3           top  
4        bottom

In [47]:
#removing redundant col[0]
cake_recipes.drop(columns = 'recipe_position', inplace=True)

In [48]:
cake_recipes.head()

225        250        275        300 recipe_name oven_position
0  61.738655  53.912627  74.414730  98.786784           a        bottom
1  51.709751  52.009735  68.576858  50.228470           a           top
2  57.095320  61.904369  61.196980  99.248541           b        bottom
3  82.455004  95.224151  98.594881  58.169349           b           top
4  96.470207  52.001358  92.893227  65.473084           c        bottom

In [57]:
#melting 'temp' columns into 1, and creating 'accuracy' col with the product
cake_melt = cake_recipes.melt(id_vars = ['recipe_name', 'oven_position'], 
                             var_name = 'temp', value_name = 'accuracy')

In [58]:
cake_melt.head()

recipe_name oven_position temp   accuracy
0           a        bottom  225  61.738655
1           a           top  225  51.709751
2           b        bottom  225  57.095320
3           b           top  225  82.455004
4           c        bottom  225  96.470207

In [93]:
#avging accuracy of ea recipe
cake_melt.groupby('recipe_name').mean()

accuracy
recipe_name           
a            63.922201
b            76.736074
c            75.874748
d            62.864844

In [59]:
#cake_tidy = cake_melt.pivot_table(index = ['recipe_name', 'oven_position'], 
                                 columns = 'temp', values = 'accuracy')

In [60]:
#cake_tidy

temp                             225        250        275        300
recipe_name oven_position                                            
a           bottom         61.738655  53.912627  74.414730  98.786784
            top            51.709751  52.009735  68.576858  50.228470
b           bottom         57.095320  61.904369  61.196980  99.248541
            top            82.455004  95.224151  98.594881  58.169349
c           bottom         96.470207  52.001358  92.893227  65.473084
            top            71.306308  82.795477  92.098049  53.960273
d           bottom         52.799753  58.670419  51.747686  56.183110
            top            96.873178  76.101363  59.571620  50.971626

In [90]:
#cake_tidy = cake_melt.pivot_table(index = ['recipe_name', 'oven_position', 'temp'])

In [165]:
##cake_tidy.reset_index(inplace = True)
#cake_tidy

3c. Which recipe, on average, is the best? recipe b

In [166]:
#avging accuracy of ea recipe
cake_melt.groupby('recipe_name').mean()

accuracy
recipe_name           
a            63.922201
b            76.736074
c            75.874748
d            62.864844

In [137]:
#pulling the recipe with the highest accuracy
cake_melt.groupby('recipe_name').mean().nlargest(1, 'accuracy')

accuracy
recipe_name           
b            76.736074

In [92]:
cake_tidy.groupby('recipe_name').mean()

accuracy
recipe_name           
a            63.922201
b            76.736074
c            75.874748
d            62.864844

In [95]:
#cake_tidy.groupby('recipe_name').mean().sum(axis=0)

3d. Which oven temperature, on average, produces the best results? 275

In [96]:
cake_melt.head(2)

recipe_name oven_position temp   accuracy
0           a        bottom  225  61.738655
1           a           top  225  51.709751

In [97]:
#avging accuracy for ea temp setting
cake_melt.groupby('temp').mean()

accuracy
temp           
225   71.306022
250   66.577437
275   74.886754
300   66.627655

In [135]:
#then pulling the temp setting with the highest accuracy
cake_melt.groupby('temp').mean().nlargest(1, 'accuracy')

accuracy
temp           
275   74.886754

3e. Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [102]:
cake_tidy = cake_melt.pivot_table(index = ['recipe_name', 'oven_position'], 
                                 columns = 'temp', values = 'accuracy')

In [103]:
cake_tidy

temp                             225        250        275        300
recipe_name oven_position                                            
a           bottom         61.738655  53.912627  74.414730  98.786784
            top            51.709751  52.009735  68.576858  50.228470
b           bottom         57.095320  61.904369  61.196980  99.248541
            top            82.455004  95.224151  98.594881  58.169349
c           bottom         96.470207  52.001358  92.893227  65.473084
            top            71.306308  82.795477  92.098049  53.960273
d           bottom         52.799753  58.670419  51.747686  56.183110
            top            96.873178  76.101363  59.571620  50.971626

In [109]:
cake_tidy.max()

temp
225    96.873178
250    95.224151
275    98.594881
300    99.248541
dtype: float64

In [110]:
cake_tidy.max().max()

99.2485405378462

In [129]:
#group data by accuracy, search for the max
#sorting the data by accuracy
cake = cake_melt.groupby('accuracy').max().sort_values(by='accuracy', ascending=False)

In [132]:
#pull the 1/highest accuracy
cake.head(1)

recipe_name oven_position temp
accuracy                                
99.248541           b        bottom  300

In [133]:
#EVEN FASTER: pulling the highest accuracy
cake_melt.nlargest(1, 'accuracy')

recipe_name oven_position temp   accuracy
26           b        bottom  300  99.248541

In [168]:
cake_melt.groupby(['temp', 'recipe_name', 'oven_position']).accuracy.mean().idxmax()

('300', 'b', 'bottom')

In [169]:
cake_melt.groupby(['temp', 'recipe_name', 'oven_position']).accuracy.mean().max()

99.2485405378462

# 4. gapminder1

In [171]:
gapminder1 = pd.read_csv('untidy-data/gapminder1.csv')
gapminder1

year      country    measure   measurement
0     1955  Afghanistan        pop  8.891209e+06
1     1960  Afghanistan        pop  9.829450e+06
2     1965  Afghanistan        pop  1.099788e+07
3     1970  Afghanistan        pop  1.243062e+07
4     1975  Afghanistan        pop  1.413202e+07
...    ...          ...        ...           ...
2074  1985    Venezuela  fertility  3.648500e+00
2075  1990    Venezuela  fertility  3.250000e+00
2076  1995    Venezuela  fertility  2.941500e+00
2077  2000    Venezuela  fertility  2.723000e+00
2078  2005    Venezuela  fertility  2.547000e+00

[2079 rows x 4 columns]

In [172]:
gapminder1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2079 entries, 0 to 2078
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   year         2079 non-null   int64  
 1   country      2079 non-null   object 
 2   measure      2079 non-null   object 
 3   measurement  2079 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 65.1+ KB


In [173]:
gapminder1.describe()

year   measurement
count  2079.000000  2.079000e+03
mean   1980.000000  1.874479e+07
std      15.815192  9.346038e+07
min    1955.000000  9.400000e-01
25%    1965.000000  5.040250e+00
50%    1980.000000  6.949800e+01
75%    1995.000000  4.537900e+06
max    2005.000000  1.303182e+09

In [175]:
gapminder1_pt = gapminder1.pivot_table(index = ['country', 'year'], columns = 'measure', values = 'measurement')
gapminder1_pt

measure           fertility  life_expect         pop
country     year                                    
Afghanistan 1955     7.7000       30.332   8891209.0
            1960     7.7000       31.997   9829450.0
            1965     7.7000       34.020  10997885.0
            1970     7.7000       36.088  12430623.0
            1975     7.7000       38.438  14132019.0
...                     ...          ...         ...
Venezuela   1985     3.6485       70.190  16997509.0
            1990     3.2500       71.150  19325222.0
            1995     2.9415       72.146  21555902.0
            2000     2.7230       72.766  23542649.0
            2005     2.5470       73.747  25375281.0

[693 rows x 3 columns]

In [178]:
gapminder1_pt.groupby('country').mean('fertility')

measure             fertility  life_expect           pop
country                                                 
Afghanistan          7.704336    38.267727  1.586984e+07
Argentina            2.926136    69.658182  2.875317e+07
Aruba                2.841909    71.410273  6.259145e+04
Australia            2.274636    75.166818  1.475517e+07
Austria              1.838182    73.676273  7.589900e+06
Bahamas              3.107291    67.793364  2.061265e+05
Bangladesh           5.046391    50.956818  9.089550e+07
Barbados             2.461818    71.504000  2.520703e+05
Belgium              1.900636    74.154636  9.756877e+06
Bolivia              5.374727    53.603727  5.632676e+06
Brazil               3.931155    63.268818  1.234202e+08
Canada               2.130364    75.462091  2.467046e+07
Chile                3.256636    68.584182  1.128061e+07
China                3.426000    63.401971  9.694329e+08
Colombia             4.201550    65.102727  2.743379e+07
Costa Rica           4.097273    71.361000  2.415921e+06
Croatia              1.846455    70.860091  4.303456e+06
Cuba                 2.591727    72.101818  9.336838e+06
Dominican Republic   4.690000    62.975091  5.778527e+06
Ecuador              4.717955    64.131364  8.254007e+06
Egypt                5.064091    57.547545  4.656314e+07
El Salvador          4.784555    60.939818  4.383817e+06
Finland              1.944727    73.577545  4.791856e+06
France               2.114755    74.979727  5.321543e+07
Georgia              2.221545    68.565545  4.742114e+06
Germany              1.683273    73.984818  7.785609e+07
Greece               1.840727    74.448909  9.460273e+06
Grenada              4.179545    65.623182  9.122109e+04
Haiti                5.384050    51.309455  5.404284e+06
Hong Kong            2.422182    74.632182  4.864401e+06
Iceland              2.628091    76.877000  2.289015e+05
India                4.487655    54.601818  7.043606e+08
Indonesia            4.037009    55.869182  1.492616e+08
Iran                 5.179636    59.888182  4.254394e+07
Iraq                 6.217455    57.605545  1.423604e+07
Ireland              2.890091    73.572455  3.329272e+06
Israel               3.311000    74.396364  3.863486e+06
Italy                1.780636    74.747818  5.482481e+07
Jamaica              3.883355    69.678364  2.151128e+06
Japan                1.712818    75.899364  1.126604e+08
Kenya                6.701364    53.627455  1.815902e+07
Lebanon              3.933364    66.769091  2.813320e+06
Mexico               4.581955    66.737909  6.901293e+07
Netherlands          2.037727    75.968364  1.387053e+07
New Zealand          2.586182    74.407636  3.122107e+06
Nigeria              6.571091    44.241091  7.350996e+07
North Korea          3.033364    66.596273  3.687435e+07
Norway               2.126091    76.131455  4.049190e+06
Pakistan             5.866755    55.922818  9.354954e+07
Peru                 4.763682    60.219091  1.766647e+07
Philippines          5.102191    62.168636  5.283192e+07
Poland               2.095091    70.983000  3.466401e+07
Portugal             2.144273    71.383455  9.625316e+06
Rwanda               7.523473    41.616273  5.167944e+06
Saudi Arabia         6.066000    60.388182  1.241186e+07
South Africa         4.464727    54.809909  3.030225e+07
South Korea          2.835555    64.839273  1.659533e+07
Spain                2.046182    75.045545  3.608779e+07
Switzerland          1.761364    76.105545  6.423138e+06
Turkey               4.162000    61.161091  4.627193e+07
United Kingdom       2.016364    74.353727  5.621204e+07
United States        2.293545    73.936545  2.293531e+08
Venezuela            4.317864    67.625455  1.525383e+07

In [183]:
gapminder1_pt.groupby('country').max().idxmax()

measure
fertility      Rwanda
life_expect     Japan
pop             China
dtype: object